### PR TITLE
Add CONTRIBUTING guidelines

### PR DESCRIPTION
**Pull request by mcornella**
## STATUS: do not merge

Repost of #2766 

---
### TODO
- [x] General structure
- [ ] FAQ page
- [x] Fill up _Problems_ and _Solutions_ sections
- [x] Write basic Pull Request needs (_i.e._ if it's a plugin, a README file)
- [ ] More information on the Volunteers page

@ncanceill @apjanke: care to weigh in?

### CODE REVIEW

**Comment by ncanceill** [comment]
Nice, this is long overdue.

Your text is well written, however I have one big issue with it: I believe the language used should be serious and straightforward. As much as I like your humorous tone, I think it is inappropriate, at least for the strict guidelines.

Maybe humour could be kept out of the strict guidelines, for instance in section names or side texts?

I will work on the wiki while we discuss this.

**Comment by apjanke** [comment]
Agreed on both parts. Thanks for putting this together! I am on deadline at the moment but will add my two cents when I am less busy mid-week.

Yeah. We should probably aim for simplified, non-idiomatic English, considering that a lot of the audience may be inexperienced devs or non-native English speakers.

Maybe links to ESR's "How to ask questions the smart way" and Spolsky's "Painless Bug Tracking" as references?

Andrew

**Comment by ncanceill** [comment]
@mcornella @apjanke I submitted a [rewrite](https://github.com/mcornella/oh-my-zsh/issues/1), what do you think?

**Comment by apjanke** [comment]
I like it.

How about changing the section headers from the declarative "You have a problem" and "You have a solution" to "If you have a problem" and "If you have a solution"? I had a bit of trouble figuring them out on first reading; I thought they were just straightforward statements of fact. Plus, the figure of speech "you have a problem" is something of an insult in American English.

Under "Submitting Pull Requests" > "Getting Started", how about adding a sentence at the very beginning like "A code contribution to oh-my-zsh can be made via a GitHub Pull Request (PR)." to introduce the "PR" abbreviation?

Under Reporting Issues > You have a problem, the list of items to include in the issue report isn't being formatted as a list. Maybe it needs a space to separate it from the main paragraph?

**Comment by vasundhar** [comment]
I would simply say " Report Bug", "Contribute"

**Comment by mcornella** [comment]
Very useful feedback, I'll get on it today after sunset.
—
Marc

**Comment by apjanke** [comment]
Just dropping a note that Neil's put together a [more technical HOWTO](https://github.com/robbyrussell/oh-my-zsh/wiki/Contribution-Technical-Practices) that discusses the nitty-gritty of getting your local repo set up to do OMZ development and PRs.

**Comment by apjanke** [comment]
Maybe we could use a different terminology than "red pill"/"blue pill"? I'm a Matrix fan, and I'm still having trouble following the distinction here. (Especially with the reference to a "Manichean" dichotomy, since these two setups are views of largely overlapping, and eventually resolved, histories, and not separate opposed worlds.) The terminology may also be off-putting to some contributors because of the association between "red pill" and the radical Men's Rights movement.

**Comment by ncanceill** [comment]
> Neil's put together a more technical HOWTO

Who's Neil? did you mean Neo? :-p

OK on the poor terminology. I [updated](https://github.com/robbyrussell/oh-my-zsh/wiki/Contribution-Technical-Practices/_compare/0da033911d47f77eda63d92a083da26a75b2bc56%5E...0da033911d47f77eda63d92a083da26a75b2bc56) the [page](https://github.com/robbyrussell/oh-my-zsh/wiki/Contribution-Technical-Practices).

**Comment by apjanke** [comment]
Oops! Sorry, somehow I seriously misread your username. :)

I like the new "simple contribution"/"maintained fork" terminology.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+your problem.
+
+If you find one, comment on it so we can know there are more people experiencing it.
+
+If not, look at the [Troubleshooting](https://github.com/robbyrussell/oh-my-zsh/wiki/Troubleshooting)
+page for instructions on how to gather data to better debug your problem.
+
+Then, you can go ahead and create an issue with as much detail as you can provide.
+It should include the data gathered as indicated above, along with:
+1. How to reproduce the problem
```
Comment (ncanceill) [inline CONTRIBUTING.md]: This apparently fails to create a proper ordered list, @mcornella can you please insert a new line before this one to fix?

Comment (mcornella) [inline CONTRIBUTING.md]: done!

**Comment by ncanceill** [comment]
Awesome! I think this is ready for merge — and you can fill up your checkboxes.

**Comment by robbyrussell** [comment]
Can I merge this now?

**Comment by mcornella** [comment]
I would like to make some adjustments first, see if I can shorten the whole thing a little bit. I'd also like confirmation from all participants that this is ok to merge. What do you guys think?

**Comment by apjanke** [comment]
+1. A few small things, below. But I think this is ok to merge, regardless of whether these items are addressed or not.

* I still think a link to ESR's "Asking Questions the Smart Way" would be a good reference to include.
* In the Bug Reports section, maybe it should use "expected behavior" instead of "correct behavior". I think this is a more common term in the field, and reminds people that bugs are about context and user expectations and use cases, not just a single objective Right behavior. 
* In the PR section, when we ask contributors to search for existing PRs before submitting their own, they should probably search the merged and closed PRs as well as just pending PRs, so they can see if the request was previously brought up and rejected, or already incorporated somewhere they might not expect.
* People posting "better" solutions to compare against existing PRs should include a reference to the existing PR in their PR.
* "e.g." is more commonly spelled with periods and not italicized.

**Comment by ncanceill** [comment]
Sorry it has been a long time.

@apjanke: respectfully, I respectively
- Disagree: we do not want to do too much hand-holding here
- Agree: see 5550ffc in https://github.com/mcornella/oh-my-zsh/pull/5
- Agree: see f556721 in https://github.com/mcornella/oh-my-zsh/pull/5
- Agree: see f556721 in https://github.com/mcornella/oh-my-zsh/pull/5
- Agree without my heart because I dislike that spelling: see 7f59db5 in https://github.com/mcornella/oh-my-zsh/pull/5

**Comment by mcornella** [comment]
I let this PR fall into the weeds, but [this feature](https://github.com/blog/2111-issue-and-pull-request-templates) will be a nice addition to the project.
Just posting for reference

**Comment by robbyrussell** [comment]
@mcornella What are our next steps here?

**Comment by belak** [comment]
Are there still plans to get this in? It seems like quite a few PRs would greatly improve in quality if there was a document with recommendations we could point people to.

**Pull request merged**